File: tests/fixtures/fixture_paymentService.js
// fixtures/paymentService.js
// Covers: function, const =, let =, async function, export function,
//         export default, class, call sites, assignments, mutations

const PROCESS_PAYMENT_TIMEOUT = 5000;

// ── Definitions ───────────────────────────────────────────────

function processPayment(order, amount) {
  const result = gateway.charge(amount);
  return processPaymentResult(result);
}

const processPaymentHandler = async (req, res) => {
  const result = await processPayment(req.body.order, req.body.amount);
  res.json(result);
};

const processPaymentValidator = function (order) {
  return order && order.total > 0;
};

async function processPaymentWithRetry(order, retries = 3) {
  for (let i = 0; i < retries; i++) {
    try {
      return await processPayment(order, order.total);
    } catch (e) {
      if (i === retries - 1) throw e;
    }
  }
}

export function processPaymentExport(order) {
  return processPayment(order, order.total);
}

export default function processPaymentDefault(order) {
  return processPayment(order, order.total);
}

class PaymentService {
  processPayment(order, amount) {
    return gateway.charge(order, amount);
  }

  async processPaymentAsync(order) {
    return await this.processPayment(order, order.total);
  }
}

// ── Call sites ────────────────────────────────────────────────

function checkout(cart) {
  const result = processPayment(cart.order, cart.total);
  return result;
}

async function handleSubmit(event) {
  event.preventDefault();
  await processPayment(currentOrder, currentOrder.amount);
}

const runBatch = (orders) => {
  return orders.map(order => processPayment(order, order.total));
};

// ── Passed as argument ────────────────────────────────────────

function withRetry(fn, times) {
  return fn();
}

const scheduled = withRetry(processPayment, 3);
queue.push(processPayment);
Promise.resolve().then(processPayment);

// ── Assignments ───────────────────────────────────────────────

let processPaymentFn = processPayment;
var processPaymentRef = processPaymentHandler;
const config = {
  processPayment: true,
  timeout: PROCESS_PAYMENT_TIMEOUT,
};
const { processPayment: payFn } = handlers;

// ── Return / resolve / emit ───────────────────────────────────

function getPaymentFn() {
  return processPayment;
}

function resolvePayment(order) {
  return new Promise((resolve) => {
    resolve(processPayment(order, order.total));
  });
}

function emitPayment(emitter, order) {
  emitter.emit("payment", processPayment(order, order.total));
}

// ── Mutations ─────────────────────────────────────────────────

const paymentQueue = [];
paymentQueue.push(processPaymentHandler);
paymentQueue.unshift(processPaymentHandler);

const paymentMap = new Map();
paymentMap.set("default", processPayment);
paymentMap.delete("old");

// ── Unrelated symbol (should NOT match) ───────────────────────

function refundPayment(order) {
  return gateway.refund(order);
}

function validateCard(card) {
  return card.number.length === 16;
}

File: tests/fixtures/fixture_payment_service.rb
# fixtures/payment_service.rb
# Covers: def, def self., class, module, lambda, proc,
#         call sites, assignments, pass-as-arg, return, yield, mutations

module Billing
  class PaymentService
    PROCESS_PAYMENT_TIMEOUT = 30

    process_payment_logger = lambda { |msg| puts msg }
    process_payment_hook   = proc { |order| order.notify }

    def initialize(gateway)
      @gateway = gateway
      @process_payment_retries = 0
    end

    # ── Definition under test ──────────────────────────────────
    def process_payment(order, amount)
      result = @gateway.charge(amount)
      return process_payment_result(result)
    end

    def self.process_payment(order)
      new(DefaultGateway).process_payment(order, order.total)
    end

    def process_payment_result(result)
      yield result if block_given?
      result
    end

    # ── Call sites ─────────────────────────────────────────────
    def retry_process_payment(order)
      @process_payment_retries += 1
      process_payment(order, order.total)
    end

    def batch_process
      orders.each do |order|
        process_payment(order, order.amount)
      end
    end

    # ── Passed as argument ─────────────────────────────────────
    def schedule(callback)
      queue.push(callback)
      run_later(:process_payment, order: @current_order)
    end

    # ── Assignments ────────────────────────────────────────────
    def build_context
      process_payment_method = :credit_card
      opts = { process_payment: true, retries: 3 }
      handler = method(:process_payment)
      [process_payment_method, opts, handler]
    end

    # ── Return / yield ─────────────────────────────────────────
    def run_payment_flow(order)
      result = process_payment(order, order.total)
      yield process_payment_result(result)
      return process_payment_result(result)
    end

    # ── Mutations ──────────────────────────────────────────────
    def reset_state
      @process_payment_retries = 0
      payment_queue.clear
      failed_payments = []
      failed_payments.push(process_payment_logger)
    end

    private

    def validate_order(order)
      raise ArgumentError, "invalid" unless order.valid?
    end
  end

  # Another class calling the service
  class OrderController
    def complete_order(order)
      PaymentService.process_payment(order)
    end

    def handle_checkout(cart)
      svc = PaymentService.new(StripeGateway)
      svc.process_payment(cart.to_order, cart.total)
    end
  end
end
